HTTPRoute Timeouts documentation (#2492)
kubernetes-sigs/gateway-api: site-src/api-types/httproute.md

Reference the [backendRef][backendRef] API documentation for additional details
on `weight` and other fields.

> #### Timeouts (optional)

Reviewer: ```suggestion
#### Timeouts
```

> 

Reviewer: ```suggestion
??? example "Experimental Channel in v1.0.0+"

    HTTPRoute timeouts are part of the Experimental Channel in `v1.0.0+`.
```

> Timeouts define API objects for L7 timeouts in HTTPRoutes. If unspecified, timeout behavior is implementation-specific.

Reviewer 2: This is the _only_ reference to "L7" throughout this document. For future readers, and maybe particularly for non-native english speakers this could be slightly confusing, would this maybe be better?

```suggestion
Timeouts resources are API objects that provide request timeouts for use in `HTTPRoutes`. If unspecified, timeout behavior is implementation-specific.
```

> Timeouts resources are API objects that provide request timeouts for use in `HTTPRoutes`. If unspecified, timeout behavior is implementation-specific.

Reviewer: ```suggestion
HTTPRoute Rules now include a `Timeouts` field. If unspecified, timeout behavior is implementation-specific.
```

> There are 2 kinds of timeouts that can be configured in an HTTPRouteRule:

Reviewer: ```suggestion
There are 2 kinds of timeouts that can be configured in an HTTPRoute Rule:
```

> 1. `request` is the timeout for the Gateway API implementation to send a response to a client HTTP request. Whether the gateway starts the timeout before or after the entire client request stream has been received, is implementation dependent. This timeout is intended to cover as close to the whole request-response transaction as possible, although an implementation MAY choose to start the timeout after the entire request stream has been received instead of immediately after the transaction is initiated by the client.

Reviewer: ```suggestion
1. `request` is the timeout for the Gateway API implementation to send a response to a client HTTP request. Whether the Gateway starts the timeout before or after the entire client request stream has been received, is implementation dependent. This timeout is intended to cover as close to the whole request-response transaction as possible, although an implementation MAY choose to start the timeout after the entire request stream has been received instead of immediately after the transaction is initiated by the client.
```

> 2. `backendRequest` is a timeout for a single request from the gateway to a backend which may be used in conjunction with retry configuration (if supported by the implementation). This timeout covers the time from when the request first starts being sent from the gateway to when the full response has been received from the backend.

Reviewer: I think I'd leave out a reference to retry configuration because it's not included in Gateway API yet so could be confusing.
```suggestion
2. `backendRequest` is a timeout for a single request from the Gateway to a backend. This timeout covers the time from when the request first starts being sent from the gateway to when the full response has been received from the backend. This can be particularly helpful if the Gateway retries connections to a backend.
```

> Because the `request` timeout encompasses the `backendRequest` timeout, the value of `backendRequest` must be less or equal to the value of `request` timeout.

Reviewer: ```suggestion
Because the `request` timeout encompasses the `backendRequest` timeout, the value of `backendRequest` must not be greater than the value of `request` timeout.
```

> Timeouts are optional, and timeouts fields they are of type `Duration`. A zero-valued timeout ("0s") is interpreted as disabling the timeout. A valid non-zero-valued timeout must be >= 1ms.

Reviewer 2: ```suggestion
Timeouts are optional and their fields are of type `Duration`. A zero-valued timeout ("0s") MUST be interpreted as disabling the timeout. A valid non-zero-valued timeout MUST be >= 1ms.
```

The following example uses the `request` field which will cause a timeout if a client request is taking longer than 10 seconds to complete. The example also defines a 2s `backendRequest` which specifies a timeout for an individual request from the gateway to a backend service `timeout-svc`:
```yaml

> {% include 'standard/http-route-timeouts/timeout-example.yaml' %}

Reviewer: Was broken link per deploy preview: https://deploy-preview-2492--kubernetes-sigs-gateway-api.netlify.app/api-types/httproute/
```suggestion
{% include 'experimental/http-route-timeouts/timeout-example.yaml' %}
```

Reference the [backendRef][backendRef] API documentation for additional details
on `weight` and other fields.

#### Timeouts (optional)

??? example "Experimental Channel in v1.0.0+"

    HTTPRoute timeouts are part of the Experimental Channel in `v1.0.0+`.

> HTTPRoute Rules now include a `Timeouts` field. If unspecified, timeout behavior is implementation-specific.

Author: I agree, I removed "now". thanks

There are 2 kinds of timeouts that can be configured in an HTTPRoute Rule:

> 1. `request` is the timeout for the Gateway API implementation to send a response to a client HTTP request. Whether the Gateway starts the timeout before or after the entire client request stream has been received, is implementation dependent. This timeout is intended to cover as close to the whole request-response transaction as possible, although an implementation MAY choose to start the timeout after the entire request stream has been received instead of immediately after the transaction is initiated by the client.

Reviewer 3: This is saying the implementation dependent part twice.

```suggestion
1. `request` is the timeout for the Gateway API implementation to send a response to a client HTTP request. This timeout is intended to cover as close to the whole request-response transaction as possible, although an implementation may choose to start the timeout after the entire request stream has been received instead of immediately after the transaction is initiated by the client.
```

2. `backendRequest` is a timeout for a single request from the Gateway to a backend. This timeout covers the time from when the request first starts being sent from the gateway to when the full response has been received from the backend. This can be particularly helpful if the Gateway retries connections to a backend.

Because the `request` timeout encompasses the `backendRequest` timeout, the value of `backendRequest` must not be greater than the value of `request` timeout.

> Timeouts are optional and their fields are of type `Duration`. A zero-valued timeout ("0s") MUST be interpreted as disabling the timeout. A valid non-zero-valued timeout MUST be >= 1ms.

Reviewer 3: Shouldn't this link to the Duration spec?

```suggestion
Timeouts are optional and their fields are of type [Duration](/geps/gep-2257/). A zero-valued timeout ("0s") MUST be interpreted as disabling the timeout. A valid non-zero-valued timeout MUST be >= 1ms.
```

Author: Thank @frankbu, maybe this should be referencing the API Specification `#gateway.networking.k8s.io/v1.Duration` instead of GEP-2257. What do you think?

Reviewer 3: Good question. The v1.Duration doc says: `The format is as specified in GEP-2257`, but nowhere is there a link to the actual spec. Should we have a link to GEP-2257 here, there, both, or neither? @robscott WDYT?

Reviewer: I don't have a strong opinion here, but tend to lean towards linking directly to the GEP since that's where they'd end up in either case.
